File: AlgorithmsPackCPP/746. Min Cost Climbing Stairs.cpp
const int MAX = 1000+1;
int memory[MAX];
vector<int>cost;

class Solution {
public:
    int climb(int index){
        if(index>=cost.size())
            return 0;
        
        auto&ret = memory[index];
        if(ret!=-1)
            return ret;
        
        int next = cost[index]+climb(index+1);
        int skip = cost[index]+climb(index+2);
        
        return ret=min(next,skip);
    }
    
    int minCostClimbingStairs(vector<int>& cost_) {
        cost = cost_;
        memset(memory,-1,sizeof(memory));
        return min(climb(0),climb(1));
    }
};

File: AlgorithmsPackPythonSolutions/746. Min Cost Climbing Stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        MAX = 1000 + 1
        memory = [-1] * MAX

        def climb(index):
            if index >= len(cost):
                return 0

            if memory[index] != -1:
                return memory[index]

            next_step = cost[index] + climb(index + 1)
            skip_step = cost[index] + climb(index + 2)

            result = min(next_step, skip_step)
            memory[index] = result
            return result

        return min(climb(0), climb(1))
